psychopy/experiment/components/_base.py

```diff
@@ -808,6 +809,29 @@ def writeFrameCodeJS(self, buff):
             buff.setIndentLevel(-1, relative=True)  # to exit the if block
             buff.writeIndented("}\n")
 
+    def writeRoutineEndCode(self, buff):
+        """Write the iohub HDF5 experiment message code that will be called
+         at the end of a routine.
+        """
+        BaseComponent.writeRoutineEndCode(self, buff)
+        saveHdf5 = self.exp.settings.params['Save hdf5 file'].val
+        if saveHdf5:
+            name = self.params['name']
+            comp_onset_tag = "%s.started" % name
+            if 'saveStartStop' in self.params and self.params['saveStartStop'].val:
+                if self.params['syncScreenRefresh'].val:
+                        code = (
+                            f"if ioServer:\n"
+                            f"    ioServer.cacheMessageEvent('{comp_onset_tag}', sec_time={name}.tStartRefresh)\n"
```

Comment:
Adding tabs to boilerplate like this can cause an indentation error in some cases, it's safer to do it like this:
```suggestion
                code = (
                        f"if ioServer:\n"
                )
                buff.writeIndentedLines(code)
                buff.setIndentLevel(1, relative=True)
                if self.params['syncScreenRefresh'].val:
                        code = (
                                f"ioServer.cacheMessageEvent('{comp_onset_tag}', sec_time={name}.tStartRefresh)\n"
                        )
                else:
                        code = (
                                f"ioServer.cacheMessageEvent('{comp_onset_tag}', sec_time={name}.tStart)\n"
                        )
                buff.writeIndentedLines(code)
                buff.setIndentLevel(-1, relative=True)
```
as `buff.setIndentLevel` will figure out how much to indent by based on user's prefs rather than assuming an indent is 4 spaces.

(though there's loads of instances where we've already done indentation wrong, particularly in older components)